### PR TITLE
Enhance startup script

### LINKED ISSUE #15663: Enhance startup script for ShardingSphere-Proxy
### Is your feature request related to a problem?

Sometimes we try to start a new Proxy process before stopping the previous. In this situation, the new Proxy will fail to startup and we won't know it until we check the log.

### Describe the feature you would like.

- Support using `$JAVA_HOME`.
- (Optional) Check if the address and port which we attempt to use in new Proxy process is already used by another Proxy.
- Consider how to join the script until the Proxy startup succeeded or failed.
- (Docker) Make container down if ShardingSphere-Proxy startup failed.

### PR DESCRIPTION
**Pull request by sunkai-cai**
Fixes #15663.

Changes proposed in this pull request:
- Refactor the script
- Priority use java from JAVA_HOME 
- import “IS_DOCKER” When run with docker

### CODE REVIEW

File: shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh [hunk 1]
```diff
@@ -137,12 +156,57 @@ if [ -z "$PORT" ]; then
...
+echo "STDOUT log file: $STDOUT_FILE"
+echo -e "Starting the $SERVER_NAME ...\c"
+
+for((i=1;i<=10;i++)); do
+  if [ "$i" = "10" ]; then
+    echo "failed: Address already in use"
+    exit
+  fi
+
+  PORT_STATUS=$(netstat -ant |$GREP $REAL_PORT |$GREP LISTEN)
```
Comment (TeslaCN) [inline shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh:178]: ShardingSphere-Proxy could listen on specified address such as `192.168.0.1/red`.

**Review by TeslaCN — CHANGES_REQUESTED** [review]

File: shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh [hunk 2]
```diff
@@ -137,12 +156,57 @@ if [ -z "$PORT" ]; then
...
+    else
+      echo "FAILED TO START"
+    fi
+  else
+    echo "FAILED TO GET PID"
+  fi
+else
+  echo "SERVER DID NOT START"
+fi
+exit 1
```
Comment (TeslaCN) [inline shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh:212]: Does it work in Docker?

Comment (sunkai-cai) [reply shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh:212]: i try it as soon

File: shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh (hunk 1 above)
Comment (sunkai-cai) [reply shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh:178]: ok, i fix it

File: shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh (hunk 2 above)
Comment (sunkai-cai) [reply shardingsphere-distribution/shardingsphere-proxy-distribution/src/main/resources/bin/start.sh:216]: ok，work in Docker is done。

**Review by TeslaCN — CHANGES_REQUESTED** [review]
I try your branch in an environment without `netstat`.

```
root@shardingsphere:/opt/shardingsphere-proxy# bin/start.sh
we find java version: java17, full_version=17.0.1, full_path=/usr/local/openjdk-17/bin/java
The classpath is /opt/shardingsphere-proxy/conf:/opt/shardingsphere-proxy/conf:.:/opt/shardingsphere-proxy/lib/*:/opt/shardingsphere-proxy/ext-lib/*
main class org.apache.shardingsphere.proxy.Bootstrap -1 /opt/shardingsphere-proxy/conf 
STDOUT log file: /opt/shardingsphere-proxy/logs/stdout.log
Starting the ShardingSphere-Proxy ...bin/start.sh: line 187: netstat: command not found
bin/start.sh: line 187: netstat: command not found
FAILED TO START
root@shardingsphere:/opt/shardingsphere-proxy# 
```

**Comment by TeslaCN** [comment]
How about echoing some warning instead of exiting if addresses & port checking failed? I'm not sure whether the checking could cover all cases.

**Comment by sunkai-cai** [comment]
> How about echoing some warning instead of exiting if addresses & port checking failed? I'm not sure whether the checking could cover all cases.

this a good idea.

**Review by TeslaCN — CHANGES_REQUESTED** [review]
I tried your branch. Proxy had already started successfully. But the script stuck.
```
17:57:33 wuweijie@wuweijie-ubuntu proxy1 enhance-startup-script- ? bin/start.sh -a 192.168.3.19,::1         
we find java version: java17, full_version=17.0.1, full_path=/usr/local/java/jdk-17.0.1/bin/java
The address is 192.168.3.19,::1
The classpath is /home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/proxy1/conf:.:/home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/proxy1/lib/*:/home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/proxy1/ext-lib/*
main class org.apache.shardingsphere.proxy.Bootstrap -1 /home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/proxy1/conf 192.168.3.19,::1
STDOUT log file: /home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/proxy1/logs/stdout.log
Starting the ShardingSphere-Proxy ....................
........................................................................................................
```

```
~/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/proxy1
☺  telnet 192.168.3.19 3307                                                                                               enhance-startup-script- b2ed7fbfea5
Trying 192.168.3.19...
Connected to 192.168.3.19.
Escape character is '^]'.
v
5.7.22-ShardingSphere-Proxy 5.1.3-SNAPSHOT-b2ed7fbwIra2bVaO�rrdhwnn7fdhymysql_native_password^C^]
telnet> Connection closed.
```

**Comment by sunkai-cai** [comment]
thanks @TeslaCN ，This is a good case。 I fix it to support  mulit-ip for start.sh

**Review by TeslaCN — CHANGES_REQUESTED** [review]
I tested the commit https://github.com/apache/shardingsphere/pull/20038/commits/5cfde330ace5e7897f6253fb2a2791694d56851c

Occupy the port by other process.
```bash
nc -lp 3307
```

Start ShardingSphere-Proxy:
```
❯ bin/start.sh                               
we find java version: java17, full_version=17.0.1, full_path=/usr/local/java/jdk-17.0.1/bin/java
The classpath is /home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/apache-shardingsphere-5.1.3-SNAPSHOT-shardingsphere-proxy-bin/conf:/home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/apache-shardingsphere-5.1.3-SNAPSHOT-shardingsphere-proxy-bin/conf:.:/home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/apache-shardingsphere-5.1.3-SNAPSHOT-shardingsphere-proxy-bin/lib/*:/home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/apache-shardingsphere-5.1.3-SNAPSHOT-shardingsphere-proxy-bin/ext-lib/*
main class org.apache.shardingsphere.proxy.Bootstrap -1 /home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/apache-shardingsphere-5.1.3-SNAPSHOT-shardingsphere-proxy-bin/conf 
STDOUT log file: /home/wuweijie/projects/shardingsphere/shardingsphere-distribution/shardingsphere-proxy-distribution/target/apache-shardingsphere-5.1.3-SNAPSHOT-shardingsphere-proxy-bin/logs/stdout.log
Starting the ShardingSphere-Proxy ............WARNING: Address already in use
. SUCCESS, PID: 4134451
```

```
Exception in thread "main" io.netty.channel.unix.Errors$NativeIoException: bind(..) failed: Address already in use
```

Proxy failed to startup, but the message show `SUCCESS`.

**Comment by TeslaCN** [comment]
There are many cases should be considered in address and port checking. Maybe we could do the checking in next PR. This PR just focus on enhancements about Docker and JAVA_HOME.

**Review by TeslaCN — APPROVED** [review]

**Pull request merged**